### PR TITLE
fix #8975 - Change how barline distance is used in order to avoid collisions in double bar lines

### LINKED ISSUE #8975: [MU4 Issue] "Barline distance" style options have unintuitive relationship with "Barline thickness" options
**Describe the bug**
In the `Barlines` section of the Style options, we have the option to change the thickness and distance of thick barlines and double barlines. However, the distance measurement measures the distance between the *centers* of the barlines, not the *edges* of the barlines.

The result of this is that *increasing* the thickness of a thick bar *reduces* the amount of separation between it and and the neighboring thin bar line.

**To Reproduce**
1. Create a new score or open an existing one
2. Go to Format > Style... > Barlines
3. Change `Thick barline thickness` to 20sp
4. Note how the double bar at the end of the piece is now malformed: the thin bar is inside of the thick bar, and the thick bar intrudes into the preceding measure(s)
5. Change `Thick barline distance` to 10.5sp
6. Note how the separation between the thick and thin barlines is now about 0.5sp

**Expected behavior**
Adjusting the thickness of a barline should not change the separation between it and neighboring barlines.
For example, if `Thick barline thickness` is 20sp and `Thick barline distance` is 0.5sp, the separation between the thick and thin barlines should be 0.5sp.

**Screenshots**
Bug:
image
Workaround:
image

**Desktop:**
 - OS: OS: Windows 10 Version 2009, Arch.: x86_64, MuseScore version (64-bit): 4.0.0-, revision: github-musescore-musescore-

### PR DESCRIPTION
**Pull request by asattely**
Resolves: https://github.com/musescore/MuseScore/issues/8975

Before, style options such as endBarDistance or doubleBarDistance described the distance between the mid points of the bar lines. This created an inconsistent distance when the line thicknesses were changed. Now, those distance style settings describe the distance between the right side of one bar and the left side of another.

image

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
Seems the vtest error should be ignored

File: src/engraving/libmscore/barline.cpp [hunk 1]
```diff
@@ -772,11 +770,11 @@ void BarLine::draw(mu::draw::Painter* painter) const
...
 
         painter->setPen(Pen(curColor(), lw, PenStyle::SolidLine, PenCapStyle::FlatCap));
-        x  += score()->styleP(Sid::endBarDistance) * mag();
+        x  += ((lw2 * .5) + score()->styleP(Sid::endBarDistance) + (lw * .5)) * mag();
         painter->drawLine(LineF(x, y1, x, y2));
 
-        x += score()->styleP(Sid::repeatBarlineDotSeparation) * mag();
-        x -= symBbox(SymId::repeatDot).width() * .5;
+        x += ((lw * .5) + score()->styleP(Sid::repeatBarlineDotSeparation)) * mag();
+        //x -= symBbox(SymId::repeatDot).width() * .5;
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/barline.cpp:777]: Please, remove this line

File: src/engraving/libmscore/barline.cpp [hunk 2]
```diff
@@ -715,11 +715,11 @@ void BarLine::draw(mu::draw::Painter* painter) const
...
         qreal lw = score()->styleP(Sid::barWidth) * mag();
         painter->setPen(Pen(curColor(), lw, PenStyle::SolidLine, PenCapStyle::FlatCap));
-        x  += score()->styleP(Sid::endBarDistance) * mag();
+        x += ((lw2 * .5) + score()->styleP(Sid::endBarDistance) + (lw * .5)) * mag();
         painter->drawLine(LineF(x, y1, x, y2));
 
-        x += score()->styleP(Sid::repeatBarlineDotSeparation) * mag();
-        x -= symBbox(SymId::repeatDot).width() * .5;
+        x += ((lw * .5) + score()->styleP(Sid::repeatBarlineDotSeparation)) * mag();
+        //x -= symBbox(SymId::repeatDot).width() * .5;
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/barline.cpp:722]: Please, remove this line

**Comment by Jojo-Schmitz** [comment]
Seems with this all double barlines are too far apart from one another? Maybe now we'd need to half `endBarDistance` and `doubleBarDistance`?

**Comment by Jojo-Schmitz** [comment]
This does show how important functioning vtests are!
My attempt to port this to 3.x and the vtests there show the problem quite clearly.

**Comment by Jojo-Schmitz** [comment]
Vs. this:
image

```
diff --git a/libmscore/barline.cpp b/libmscore/barline.cpp
index b6d96eb162..c872d77da4 100644
--- a/libmscore/barline.cpp
+++ b/libmscore/barline.cpp
@@ -585,7 +585,7 @@ void BarLine::draw(QPainter* painter) const
 
                   qreal lw2 = score()->styleP(Sid::endBarWidth) * mag();
                   painter->setPen(QPen(curColor(), lw2, Qt::SolidLine, Qt::FlatCap));
-                  x += ((lw * .5) + score()->styleP(Sid::endBarDistance) + (lw2 * .5)) * mag();
+                  x += ((lw * .5) + (score()->styleP(Sid::endBarDistance) * .5) + (lw2 * .5)) * mag();
                   painter->drawLine(QLineF(x, y1, x, y2));
                   }
                   break;
@@ -595,7 +595,7 @@ void BarLine::draw(QPainter* painter) const
                   painter->setPen(QPen(curColor(), lw, Qt::SolidLine, Qt::FlatCap));
                   qreal x = lw * .5;
                   painter->drawLine(QLineF(x, y1, x, y2));
-                  x += ((lw * .5) + score()->styleP(Sid::doubleBarDistance) + (lw * .5)) * mag();
+                  x += ((lw * .5) + (score()->styleP(Sid::doubleBarDistance) * .5) + (lw * .5)) * mag();
                   painter->drawLine(QLineF(x, y1, x, y2));
                   }
                   break;
@@ -608,7 +608,7 @@ void BarLine::draw(QPainter* painter) const
 
                   qreal lw2 = score()->styleP(Sid::barWidth) * mag();
                   painter->setPen(QPen(curColor(), lw2, Qt::SolidLine, Qt::FlatCap));
-                  x += ((lw * .5) + score()->styleP(Sid::endBarDistance) + (lw2 * .5)) * mag();
+                  x += ((lw * .5) + (score()->styleP(Sid::endBarDistance) * .5) + (lw2 * .5)) * mag();
                   painter->drawLine(QLineF(x, y1, x, y2));
                   }
                   break;
@@ -625,7 +625,7 @@ void BarLine::draw(QPainter* painter) const
                   painter->setPen(QPen(curColor(), lw2, Qt::SolidLine, Qt::FlatCap));
                   qreal x = lw2 * .5;
                   painter->drawLine(QLineF(x, y1, x, y2));
-                  x += ((lw2 * .5) + score()->styleP(Sid::endBarDistance) + (lw2 * .5)) * mag();
+                  x += ((lw2 * .5) + (score()->styleP(Sid::endBarDistance) * .5) + (lw2 * .5)) * mag();
                   painter->drawLine(QLineF(x, y1, x, y2));
                   }
                   break;
@@ -638,7 +638,7 @@ void BarLine::draw(QPainter* painter) const
 
                   qreal lw = score()->styleP(Sid::barWidth) * mag();
                   painter->setPen(QPen(curColor(), lw, Qt::SolidLine, Qt::FlatCap));
-                  x += ((lw2 * .5) + score()->styleP(Sid::endBarDistance) + (lw * .5)) * mag();
+                  x += ((lw2 * .5) + (score()->styleP(Sid::endBarDistance) * .5) + (lw * .5)) * mag();
                   painter->drawLine(QLineF(x, y1, x, y2));
 
                   x += ((lw * .5) + score()->styleP(Sid::repeatBarlineDotSeparation)) * mag();
@@ -661,7 +661,7 @@ void BarLine::draw(QPainter* painter) const
                   painter->drawLine(QLineF(x, y1, x, y2));
 
                   qreal lw2 = score()->styleP(Sid::endBarWidth) * mag();
-                  x += ((lw * .5) + score()->styleP(Sid::endBarDistance) + (lw2 * .5)) * mag();
+                  x += ((lw * .5) + (score()->styleP(Sid::endBarDistance) * .5) + (lw2 * .5)) * mag();
                   painter->setPen(QPen(curColor(), lw2, Qt::SolidLine, Qt::FlatCap));
                   painter->drawLine(QLineF(x, y1, x, y2));
 
@@ -681,7 +681,7 @@ void BarLine::draw(QPainter* painter) const
                   painter->drawLine(QLineF(x, y1, x, y2));
 
                   qreal lw2 = score()->styleP(Sid::endBarWidth) * mag();
-                  x += ((lw * .5) + score()->styleP(Sid::endBarDistance) + (lw2 * .5)) * mag();
+                  x += ((lw * .5) + (score()->styleP(Sid::endBarDistance) * .5) + (lw2 * .5)) * mag();
                   painter->setPen(QPen(curColor(), lw2, Qt::SolidLine, Qt::FlatCap));
                   painter->drawLine(QLineF(x, y1, x, y2));
 
@@ -689,7 +689,7 @@ void BarLine::draw(QPainter* painter) const
                         drawTips(painter, true, x + lw2 * .5);
 
                   painter->setPen(QPen(curColor(), lw, Qt::SolidLine, Qt::FlatCap));
-                  x  += ((lw2 * .5) + score()->styleP(Sid::endBarDistance) + (lw * .5)) * mag();
+                  x  += ((lw2 * .5) + (score()->styleP(Sid::endBarDistance) * .5) + (lw * .5)) * mag();
                   painter->drawLine(QLineF(x, y1, x, y2));
 
                   x += ((lw * .5) + score()->styleP(Sid::repeatBarlineDotSeparation)) * mag();
@@ -1244,16 +1244,16 @@ qreal BarLine::layoutWidth(Score* score, BarLineType type)
       switch (type) {
             case BarLineType::DOUBLE:
                   w = (score->styleP(Sid::doubleBarWidth) * 2)
-                     + score->styleP(Sid::doubleBarDistance);
+                     + (score->styleP(Sid::doubleBarDistance) * .5);
                   break;
             case BarLineType::DOUBLE_HEAVY:
                   w = (score->styleP(Sid::endBarWidth) * 2)
-                     + score->styleP(Sid::endBarDistance);
+                     + (score->styleP(Sid::endBarDistance) * .5);
                   break;
             case BarLineType::END_START_REPEAT:
                   w = score->styleP(Sid::endBarWidth)
                      + (score->styleP(Sid::barWidth) * 2)
-                     + (score->styleP(Sid::endBarDistance) * 2)
+                     + score->styleP(Sid::endBarDistance)
                      + (score->styleP(Sid::repeatBarlineDotSeparation) * 2)
                      + (dotwidth * 2);
                   break;
@@ -1261,7 +1261,7 @@ qreal BarLine::layoutWidth(Score* score, BarLineType type)
             case BarLineType::END_REPEAT:
                   w = score->styleP(Sid::endBarWidth)
                      + score->styleP(Sid::barWidth)
-                     + score->styleP(Sid::endBarDistance)
+                     + (score->styleP(Sid::endBarDistance) * .5)
                      + score->styleP(Sid::repeatBarlineDotSeparation)
                      + dotwidth;
                   break;
@@ -1269,7 +1269,7 @@ qreal BarLine::layoutWidth(Score* score, BarLineType type)
             case BarLineType::REVERSE_END:
                   w = score->styleP(Sid::endBarWidth)
                      + score->styleP(Sid::barWidth)
-                     + score->styleP(Sid::endBarDistance);
+                     + (score->styleP(Sid::endBarDistance) * .5);
                   break;
             case BarLineType::BROKEN:
             case BarLineType::NORMAL:
```

**Comment by asattely** [comment]
The bars are now too far apart because the 'distance' style parameters are now interpreted in a different way. I'm going to add a commit today that changes those default values to bring them closer to how it looks like in previous versions. Stay tuned!

**Comment by Jojo-Schmitz** [comment]
That though would influence how these settings from scores imported from 3.x do get interpreted

**Comment by asattely** [comment]
Removed unnecessary commented code and changed the defaults in styledef to account for distance interpretation. Assuming all stock default options, I subtracted the amount of space that was eaten up by line thickness (for example double bar distance I subtracted the default double bar thickness because half of one barline and half of the other used to be included in the distance value)

**Comment by Jojo-Schmitz** [comment]
As said: changing those defaults will fix the issue for new scores and for imported (pre 4.x) scores that never set those valued away from the default, but not for those that did change those values. So on import, these need to get adjusted.
Easiest might be to just half them?

**Comment by asattely** [comment]
> As said: changing those defaults will fix the issue for new scores and for imported (pre 4.x) scores that never set those valued away from the default, but not for those that did change those values. So on import, these need to get adjusted.
> Easiest might be to just half them?

The process of converting old scores to new scores is going to be separate functionality--the current style defaults are the ones that apply to new scores in 4.0. How we deal with upconverting 3.x scores is going to be a different issue we'll see in the future, and those decisions can be made at that point I figure

**Comment by Jojo-Schmitz** [comment]
IMHO these need to be decided upon now, else it'd get forgotten.

**Comment by asattely** [comment]
Well I'm in the process of writing up a big list of added and changed style parameters that can be maintained by whoever is changing or adding them, with old and new defaults, but in any case old score conversion is outside the scope of this particular issue. Conversion is going to be a big issue that tackles a lot of different aspects and I think it'd be better to have all of the work for that centralized in one location rather than changed piecemeal in dozens of other pull requests.

Certainly the code that is changed in this pull request has absolutely nothing to do with score conversion, so for now let's keep this the way it is and I'll post the link to the wiki page I create today later.

**Comment by Jojo-Schmitz** [comment]
Well, as far as I'm aware these 3 values are the first and only changed defaults.

**Comment by asattely** [comment]
https://github.com/musescore/MuseScore/wiki/Style-Parameter-changes-between-3.6-and-4.0

**Comment by DmitryArefiev** [comment]
Checked #8975 on Win10, Mac11.13.1, Linux Mint 20.2 - FIXED

Please merge into master (@RomanPudashkin)

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**